kms/tpmkms/tpmkms.go

```diff
@@ -503,6 +618,84 @@ func (k *TPMKMS) LoadCertificateChain(req *apiv1.LoadCertificateChainRequest) ([
 	return chain, nil
 }
 
+const (
+	// maximumIterations is the maximum number of times for the recursive
+	// intermediate CA lookup loop.
+	maximumIterations = 10
+)
+
+func (k *TPMKMS) loadCertificateChainFromWindowsCertificateStore(req *apiv1.LoadCertificateRequest) ([]*x509.Certificate, error) {
+	pub, err := k.GetPublicKey(&apiv1.GetPublicKeyRequest{
+		Name: req.Name,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed retrieving public key: %w", err)
+	}
+
+	o, err := parseNameURI(req.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed parsing %q: %w", req.Name, err)
+	}
+
+	location := k.windowsCertificateStoreLocation
+	if o.storeLocation != "" {
+		location = o.storeLocation
+	}
+	store := k.windowsCertificateStore
+	if o.store != "" {
+		store = o.store
+	}
+
+	subjectKeyID, err := generateWindowsSubjectKeyID(pub)
+	if err != nil {
+		return nil, fmt.Errorf("failed generating subject key id: %w", err)
+	}
+
+	cert, err := k.windowsCertificateManager.LoadCertificate(&apiv1.LoadCertificateRequest{
+		Name: fmt.Sprintf("capi:key-id=%s;store-location=%s;store=%s;", subjectKeyID, location, store),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed retrieving certificate using Windows platform cryptography provider: %w", err)
+	}
+
+	intermediateCAStoreLocation := k.windowsIntermediateStoreLocation
+	if o.intermediateStoreLocation != "" {
+		intermediateCAStoreLocation = o.intermediateStoreLocation
+	}
+
+	intermediateCAStore := k.windowsIntermediateStore
+	if o.intermediateStore != "" {
+		intermediateCAStore = o.intermediateStore
+	}
+
+	chain := []*x509.Certificate{cert}
+	child := cert
+	for i := 0; i < maximumIterations; i++ { // loop a maximum of times
+		authorityKeyID := hex.EncodeToString(child.AuthorityKeyId)
+		parent, err := k.loadIntermediateFromWindowsCertificateStore(authorityKeyID, intermediateCAStoreLocation, intermediateCAStore)
+		if err != nil {
+			if errors.Is(err, apiv1.NotFoundError{}) {
+				// if error indicates the parent wasn't found, assume end of chain for a specific
+				// combination of store location and store is reached, and break from the loop
+				break
+			}
+			return nil, fmt.Errorf("failed loading intermediate CA certificate using Windows platform cryptography provider: %w", err)
+		}
+		chain = append(chain, parent)
+		child = parent
+	}
```

Comment:
Technically the chain could contain roots using this logic, if a root was placed in the "CA" certificate store. However, the "CA" certificate store is intended for intermediate CA certificates, so it's not likely to happen, and certainly not for certificate chains that we manage in our applications. It's possible to override the "CA" store with "Root" when configuring the KMS, so that way intermediates could get mixed with roots. Want me to add a check for root certs, and exclude these explicitly?

Reply:
I've added a rudimentary check for root CAs in a4b14673e61ab84c9a36ddaed531fc3706e7ac08. It'll only work in the case signature validation succeeds, but assuming an invalid signature or certificate contents are caught earlier, I think it's sufficient (for now).